.github/workflows/build.yml

```diff
@@ -1,53 +1,54 @@
 name: Build
 
 on:
+  workflow_dispatch:
   push:
     branches: [main]
   pull_request:
     branches: [main]
 
 env:
   ARM_GNU_TOOLCHAIN_VERSION: 14.2.rel1
+  ARM_GNU_TOOLCHAIN_PLATFORM: x86_64-arm-none-eabi
   EM_VERSION: 4.0.1
-  EM_CACHE_FOLDER: 'emsdk-cache'
+  EM_NODE_VERSION: 20.18.0_64bit
 
 jobs:
   build:
     runs-on: ubuntu-24.04
+    permissions:
+      packages: read
     strategy:
       matrix:
         board: ["snowy_bb2", "spalding_bb2", "silk_bb2", "asterix_evt1"]
+    container:
+      image: ghcr.io/${{ github.repository }}/builder:latest
```

Comment:
using `latest` is dangerous/problematic. which image will this pick? the latest uploaded from a merge to main?

Reply:
I put the condition here, put we should use `pull_request_target` to eliminate the way people can override it.

```yml
push: ${{ github.event_name != 'pull_request' }}
```

Reply:
so from what I understand, this will pick the image which was built on the latest commit to `main`, right? If in a PR people modify e.g. requirements.txt, their CI will fail.

Reply:
@tdtgit ping

Reply:
Sorry I'm thinking how we approach here 🤔. Yes your assumption is correct. Another case is Docker build and FW build job is run at the same time, so FW build actually using the staled Docker image.

<img width="395" alt="image" src="https://github.com/user-attachments/assets/f2708874-9546-4315-aba5-fc4d51ecf0e9" />

Caching key can work, but will made the submodules stale. Maybe we can download the submodule as a zip file, like the approach when we load `openocd`'s submodules.